src/main/java/com/urbanairship/datacube/OffsetBasedAddressFormat.java

```diff
@@ -0,0 +1,252 @@
+            }
+        }
+
+
+        if (bucket.length != dimension.getNumFieldBytes()) {
+            throw new IllegalArgumentException("Field length was wrong (after bucketing " +
+                    " and unique ID substitution). For dimension " + dimension +
+                    ", expected length " + dimension.getNumFieldBytes() + " but was " +
+                    bucket.length);
+        }
+        return bucket;
+    }
+
+    @Override
+    public Optional<Address> fromKey(byte[] bytes) throws IOException, InterruptedException {
+        int remainingBytes = bytes.length;
+
+        ByteBuffer bb = ByteBuffer.wrap(bytes);
+
+        if (useAddressPrefixByteHash) {
+            // then read the first int and throw it out.
+            byte hashByte = bb.get();
+            remainingBytes -= Ints.BYTES;
+            Preconditions.checkState(hashByte == getHashByte(bytes));
+        }
+
+
+        Address address = addressSupplier.apply(useAddressPrefixByteHash, dimensions);
+
+        for (Dimension<?, ?> dimension : dimensions) {
+            if (remainingBytes <= 0) {
+                // we might exit early if it turns out that we have nothing but wildcards remaining, which are omitted
+                // from the serialized format.
+                return Optional.of(address);
+            }
+
+            byte[] wildcardOrNah = new byte[1];
+            byte[] bucketPrefix = new byte[dimension.getBucketPrefixSize()];
+            byte[] bucket = new byte[dimension.getNumFieldBytes()];
+
+            bb.get(wildcardOrNah);
+            remainingBytes -= wildcardOrNah.length;
+
+            if (Arrays.equals(wildcardOrNah, WILDCARD_FIELD)) {
+                // then we need to consume dimension bytes, but do nothing else.
+                bb.get(bucket);
+                remainingBytes -= bucket.length;
+                if (remainingBytes <= 0) {
+                    // we might exit early if it turns out that we have nothing but wildcards remaining, which are omitted
+                    // from the serialized format.
+                    return Optional.of(address);
+                }
+                // nothing to add for the address for the dimension.
+                continue;
+            }
+
+            bb.get(bucketPrefix);
+            remainingBytes -= bucketPrefix.length;
+
+            bb.get(bucket);
+            remainingBytes -= bucket.length;
+
+            if (dimension.getDoIdSubstitution()) {
+                Optional<byte[]> maybeBucket = idService.getValueForId(dimensions.indexOf(dimension), bucket);
+                if (!maybeBucket.isPresent()) {
+                    return Optional.empty();
+                }
+                bucket = maybeBucket.get();
+            }
+
+            BucketType bucketType = null;
+            for (BucketType candidate : dimension.getBucketer().getBucketTypes()) {
+                if (Arrays.equals(candidate.getUniqueId(), bucketPrefix)) {
```

Comment:
this strikes me as a wild ride, but I can't actually come up with a problem with it.

Reply:
what's wild about it? how could i make it less wild?

Reply:
Can't think of a great way. I guess maybe move it into Bucketer would be my only idea? Anyway it's fine, I'm just being a jerk.